### ISSUE TITLE
ptr_arg with multiple `&mut Vec<u8>` sometimes has false positives.

### ISSUE THREAD

**Issue by jorgecarleitao** [rust-lang/rust-clippy#9067]
### Summary

Passing multiple `&mut Vec<u8>` and using Vec-only functions sometimes causes the linter to incorrectly warn that a slice could be passed.

things like 

```
    a: &mut Vec<u8>,
    b: &mut Vec<u8>,
    c: &mut Vec<u8>,
```

will warn on `b` and `c` but not `a` (despite every one of them being required as Vec).

### Lint Name

ptr_arg

### Reproducer

I tried this code:

```rust
pub fn read_batch<R: Read + Seek>(
    reader: &mut R,
    dictionaries: &Dictionaries,
    metadata: &FileMetadata,
    projection: Option<&[usize]>,
    index: usize,
    message_scratch: &mut Vec<u8>,
    data_scratch: &mut Vec<u8>,
) -> Result<Chunk<Box<dyn Array>>> {
    let block = metadata.blocks[index];

    let offset: u64 = block
        .offset
        .try_into()
        .map_err(|_| Error::from(OutOfSpecKind::NegativeFooterLength))?;

    // read length
    reader.seek(SeekFrom::Start(offset))?;
    let mut meta_buf = [0; 4];
    reader.read_exact(&mut meta_buf)?;
    if meta_buf == CONTINUATION_MARKER {
        // continuation marker encountered, read message next
        reader.read_exact(&mut meta_buf)?;
    }
    let meta_len = i32::from_le_bytes(meta_buf)
        .try_into()
        .map_err(|_| Error::from(OutOfSpecKind::UnexpectedNegativeInteger))?;

    message_scratch.clear();
    message_scratch.try_reserve(meta_len)?;
    reader
        .by_ref()
        .take(meta_len as u64)
        .read_to_end(message_scratch)?;

    let message = arrow_format::ipc::MessageRef::read_as_root(message_scratch.as_ref())
        .map_err(|err| Error::from(OutOfSpecKind::InvalidFlatbufferMessage(err)))?;

    let batch = get_serialized_batch(&message)?;

    let offset: u64 = block
        .offset
        .try_into()
        .map_err(|_| Error::from(OutOfSpecKind::NegativeFooterLength))?;

    let length: u64 = block
        .meta_data_length
        .try_into()
        .map_err(|_| Error::from(OutOfSpecKind::NegativeFooterLength))?;

    read_record_batch(
        batch,
        &metadata.schema.fields,
        &metadata.ipc_schema,
        projection,
        dictionaries,
        message
            .version()
            .map_err(|err| Error::from(OutOfSpecKind::InvalidFlatbufferVersion(err)))?,
        reader,
        offset + length,
        metadata.size,
        data_scratch,
    )
}
```

I saw this happen:

<img width="593" alt="Screenshot 2022-06-29 at 06 47 11" src="https://user-images.githubusercontent.com/2772607/176452405-e4a61fea-421d-4162-bc28-70909d62fcc2.png">

I expected to not see the warning, since there is a 
```
message_scratch.clear();
message_scratch.try_reserve(meta_len)?;
```

on it which is only applicable to `Vec`

### Version

```text
rustc 1.60.0 (7737e0b5c 2022-04-04)
binary: rustc
commit-hash: 7737e0b5c4103216d6fd8cf941b7ab9bdbaace7c
commit-date: 2022-04-04
host: x86_64-unknown-linux-gnu
release: 1.60.0
LLVM version: 14.0.0
```

**Comment by jorgecarleitao** [comment]
> This should be working in `1.61`

confirmed. Sorry for the noise :/
